vllm/model_executor/model_optimizer/fused_op_generator_utils.py

```diff
@@ -0,0 +1,152 @@
+    the associated function schema.
+    """
+    s = get_function_schema(n)
+    if not s or arg_num >= len(s.arguments):
+        return False
+    return isinstance(s.arguments[arg_num].real_type, torch._C.OptionalType)
+
+
+def build_extension(lib_name: str,
+                    sources: List[str],
+                    opt: str = '-O2',
+                    extra_cflags: Optional[List[str]] = None,
+                    extra_ldflags: Optional[List[str]] = None,
+                    verbose: bool = False):
+    """
+    Given a list of cpp and cuda source files, build and load a pytorch
+    extension module with the given name.  Loaded ops will appear in the
+    torch.ops.{lib_name} namespace.
+    """
+    if not extra_cflags:
+        extra_cflags = []
+
+    if not extra_ldflags:
+        extra_ldflags = []
+
+    torch.utils.cpp_extension.load(
+        name=lib_name,
+        sources=sources,
+        extra_cflags=[opt, f'-DLIBRARY_NAME={lib_name}', *extra_cflags],
+        extra_ldflags=extra_ldflags,
+        verbose=verbose,
+        is_python_module=False,
+    )
+
+
+def arg_schema_type(n: torch.fx.node.Argument,
+                    add_prefix: bool = False) -> str:
+    """
+    Get the schema or C++ type for a fused op argument.
+    """
+    if isinstance(n, float):
+        return "float"
+    elif isinstance(n, int):
+        return "int"
+    else:
+        if n.type is not None:
+            ty = n.type.__name__
+        elif n.meta.get(
+                'type') and n.meta.get('type').__name__ != 'FakeTensor':
+            ty = n.meta.get('type').__name__
+            if ty == 'Size':
+                return 'std::vector<int64_t> const' if add_prefix else 'int[]'
+        else:
+            # this default is a bit sketchy
+            ty = "Tensor"
+
+    builtin_types = {"int": "int64_t", "float": "double"}
+
+    if add_prefix and ty in builtin_types:
+        return builtin_types[ty]
+
+    return ty if not add_prefix else f"torch::{ty}"
+
+
+def generate_op_schema(
+        inputs: Dict[str, torch.fx.node.Argument],
+        outputs: List[torch.fx.Node], nodes: List[torch.fx.Node],
+        kwargs: Dict[str, Dict[str, torch.fx.node.Argument]]) -> str:
+    sep = "("
+    arg_sig = ""
+    for name in inputs:
+        arg_type = arg_schema_type(inputs[name]).replace(".", "::")
+        arg_name = inputs[name].name.replace(".", "_")
+        arg_sig = arg_sig + sep + f"{arg_type} {arg_name}"
+        sep = ", "
+
+    # TODO support kwargs
+    assert len(kwargs) == 0
+
+    arg_sig = arg_sig + ") -> "
+
+    sep = "(" if len(outputs) != 1 else ""
+
+    for i, n in enumerate(outputs):
+        arg_type = arg_schema_type(n).replace(".", "::")
+        arg_sig = arg_sig + sep + arg_type
+        sep = ", "
+
+    if len(outputs) != 1:
+        arg_sig = arg_sig + ")"
+
+    return arg_sig
+
+
+def generate_meta_function(nodes: List[torch.fx.Node]) -> Callable:
```

Comment:
I think the signature of this function should remain as it was before so that we could do a proper job of generating a meta function at some point in the future.  The `compose` stuff doesn't really work for non-unary meta functions.